### PR TITLE
`OrderedColumnLeaf` work

### PR DESCRIPTION
**Pull request by Kixiron**
* Switches `OrdKeyBatch` and `OrdValBatch` to use `OrderedColumnLeaf`
* Adds some `OrderedColumnLeaf` benchmarks
* Switches `UnorderedColumnLeafBuilder` to use columns of values instead of a `Vec<(K, R)>`

### CODE REVIEW

File: src/trace/consolidation.rs [hunk 1]
```diff
@@ -86,24 +142,149 @@ where
...
+
+    // Clear and pre-allocate the indices buffer
+    indices.clear();
+    indices.reserve(keys.len());
+    // TODO: We can do this in a vectorized manner, the assembly isn't ideal https://godbolt.org/z/4TbK6Mzec
+    indices.extend(0..keys.len());
+
+    // Ideally we'd combine the sorting and value merging portions
+    // This line right here is literally the hottest code within the entirety of the
+    // program. It makes up 90% of the work done while joining or merging anything
```
Comment (ryzhyk) [inline src/trace/consolidation.rs:173]: This makes it three hottest lines :)

**Review by ryzhyk — COMMENTED** [review]
Looks good. Waiting for @Kixiron to add perf numbers to the PR.  Do we have a way to proptest this?

**Comment by Kixiron** [comment]
Got perf numbers from the current master, this PR introduces almost unilateral perf gains apart from the cost of pushing small (<100k) numbers of elements which is to be expected since vector growth has a much larger impact on smaller cases until it gets amortized away

```
add/0
time:   [37.659 ns 37.943 ns 38.092 ns]

add/10
time:   [255.41 ns 257.54 ns 260.40 ns]

add/100
time:   [858.30 ns 860.81 ns 863.43 ns]

add/1000
time:   [7.9732 µs 8.0161 µs 8.0566 µs]

add/10,000
time:   [83.199 µs 83.824 µs 84.269 µs]

add/100,000
time:   [874.96 µs 879.34 µs 884.38 µs]

add/1,000,000
time:   [11.627 ms 11.837 ms 12.131 ms]

add/10,000,000
time:   [104.80 ms 105.58 ms 106.68 ms]

add/100,000,000
time:   [1.0393 s 1.0521 s 1.0676 s]
```

```
add-by-ref/0
time:   [45.870 ns 45.941 ns 45.992 ns]

add-by-ref/10
time:   [166.67 ns 167.35 ns 167.71 ns]

add-by-ref/100
time:   [765.37 ns 770.40 ns 775.25 ns]

add-by-ref/1000
time:   [7.7521 µs 7.7743 µs 7.8039 µs]

add-by-ref/10,000
time:   [81.201 µs 81.396 µs 81.561 µs]

add-by-ref/100,000
time:   [824.76 µs 829.05 µs 832.64 µs]

add-by-ref/1,000,000
time:   [9.5393 ms 9.5919 ms 9.6416 ms]

add-by-ref/10,000,000
time:   [93.050 ms 93.826 ms 95.101 ms]

add-by-ref/100,000,000
time:   [958.83 ms 970.60 ms 987.32 ms]
```

```
add-assign/0
time:   [37.044 ns 37.197 ns 37.354 ns]

add-assign/10
time:   [268.54 ns 271.34 ns 274.43 ns]

add-assign/100
time:   [884.43 ns 890.04 ns 894.91 ns]

add-assign/1000
time:   [8.0023 µs 8.0406 µs 8.0906 µs]

add-assign/10,000
time:   [83.438 µs 84.199 µs 84.828 µs]

add-assign/100,000
time:   [832.34 µs 834.47 µs 838.66 µs]

add-assign/1,000,000
time:   [10.859 ms 10.898 ms 10.930 ms]

add-assign/10,000,000
time:   [105.32 ms 105.95 ms 106.30 ms]

add-assign/100,000,000
time:   [1.1678 s 1.2033 s 1.2482 s]
```

```
add-assign-by-ref/0
time:   [36.253 ns 36.306 ns 36.369 ns]

add-assign-by-ref/10
time:   [243.65 ns 249.73 ns 254.79 ns]

add-assign-by-ref/100
time:   [850.63 ns 855.76 ns 860.06 ns]

add-assign-by-ref/1000
time:   [7.9372 µs 7.9703 µs 8.0069 µs]

add-assign-by-ref/10,000
time:   [82.812 µs 83.071 µs 83.362 µs]

add-assign-by-ref/100,000
time:   [822.95 µs 827.49 µs 832.29 µs]

add-assign-by-ref/1,000,000
time:   [10.726 ms 10.796 ms 10.843 ms]

add-assign-by-ref/10,000,000
time:   [104.16 ms 105.33 ms 106.38 ms]

add-assign-by-ref/100,000,000
time:   [1.1058 s 1.1187 s 1.1316 s]
```

```
neg/0
time:   [35.936 ns 36.042 ns 36.142 ns]

neg/10
time:   [37.274 ns 37.386 ns 37.464 ns]

neg/100
time:   [46.009 ns 46.082 ns 46.174 ns]

neg/1000
time:   [171.26 ns 171.50 ns 171.77 ns]

neg/10,000
time:   [1.4330 µs 1.4356 µs 1.4397 µs]

neg/100,000
time:   [14.805 µs 14.899 µs 15.051 µs]

neg/1,000,000
time:   [358.93 µs 363.07 µs 370.51 µs]

neg/10,000,000
time:   [4.9827 ms 5.0703 ms 5.1756 ms]

neg/100,000,000
time:   [55.118 ms 56.504 ms 58.263 ms]
```

```
neg-by-ref/0
time:   [45.223 ns 45.260 ns 45.311 ns]

neg-by-ref/10
time:   [98.202 ns 98.909 ns 99.816 ns]

neg-by-ref/100
time:   [116.58 ns 118.19 ns 119.82 ns]

neg-by-ref/1000
time:   [376.14 ns 391.80 ns 407.49 ns]

neg-by-ref/10,000
time:   [3.8534 µs 3.8741 µs 3.8935 µs]

neg-by-ref/100,000
time:   [36.027 µs 36.319 µs 36.658 µs]

neg-by-ref/1,000,000
time:   [2.1129 ms 2.1295 ms 2.1409 ms]

neg-by-ref/10,000,000
time:   [22.470 ms 22.636 ms 22.792 ms]

neg-by-ref/100,000,000
time:   [245.92 ms 249.27 ms 253.47 ms]
```

```
build-unordered-pushing/0
time:   [35.889 ns 35.951 ns 36.014 ns]

build-unordered-pushing/10
time:   [272.04 ns 273.64 ns 275.20 ns]

build-unordered-pushing/100
time:   [601.11 ns 602.92 ns 605.23 ns]

build-unordered-pushing/1000
time:   [2.5626 µs 2.5668 µs 2.5712 µs]

build-unordered-pushing/10,000
time:   [22.226 µs 22.901 µs 23.525 µs]

build-unordered-pushing/100,000
time:   [594.64 µs 603.66 µs 614.46 µs]

build-unordered-pushing/1,000,000
time:   [6.9153 ms 6.9397 ms 6.9642 ms]

build-unordered-pushing/10,000,000
time:   [89.800 ms 90.445 ms 91.134 ms]

build-unordered-pushing/100,000,000
time:   [1.0232 s 1.0371 s 1.0514 s]
```

```
build-unordered-boundary/0
time:   [38.582 ns 38.697 ns 38.769 ns]

build-unordered-boundary/10
time:   [92.162 ns 92.424 ns 92.834 ns]

build-unordered-boundary/100
time:   [905.29 ns 917.84 ns 927.16 ns]

build-unordered-boundary/1000
time:   [11.374 µs 11.413 µs 11.474 µs]

build-unordered-boundary/10,000
time:   [225.32 µs 226.72 µs 227.52 µs]

build-unordered-boundary/100,000
time:   [2.7121 ms 2.7674 ms 2.8007 ms]

build-unordered-boundary/1,000,000
time:   [33.926 ms 34.245 ms 34.503 ms]

build-unordered-boundary/10,000,000
time:   [387.22 ms 391.93 ms 396.67 ms]

build-unordered-boundary/100,000,000
time:   [4.5041 s 4.5539 s 4.6049 s]
```

```
build-unordered-done/0
time:   [51.940 ns 52.044 ns 52.137 ns]

build-unordered-done/10
time:   [187.17 ns 187.98 ns 189.40 ns]

build-unordered-done/100
time:   [1.1537 µs 1.1577 µs 1.1647 µs]

build-unordered-done/1000
time:   [11.765 µs 11.812 µs 11.843 µs]

build-unordered-done/10,000
time:   [234.90 µs 235.76 µs 236.69 µs]

build-unordered-done/100,000
time:   [2.9112 ms 2.9337 ms 2.9624 ms]

build-unordered-done/1,000,000
time:   [36.135 ms 36.391 ms 36.547 ms]

build-unordered-done/10,000,000
time:   [428.33 ms 434.04 ms 439.59 ms]

build-unordered-done/100,000,000
time:   [4.8069 s 4.8641 s 4.9200 s]
```

```
ordered-builder-push-merge/0
time:   [39.689 ns 39.760 ns 39.842 ns]

ordered-builder-push-merge/10
time:   [232.10 ns 232.62 ns 233.22 ns]

ordered-builder-push-merge/100
time:   [870.33 ns 871.52 ns 872.90 ns]

ordered-builder-push-merge/1000
time:   [8.2505 µs 8.2661 µs 8.2837 µs]

ordered-builder-push-merge/10,000
time:   [86.577 µs 86.764 µs 86.997 µs]

ordered-builder-push-merge/100,000
time:   [862.92 µs 864.43 µs 865.81 µs]

ordered-builder-push-merge/1,000,000
time:   [10.682 ms 10.711 ms 10.742 ms]

ordered-builder-push-merge/10,000,000
time:   [105.69 ms 106.14 ms 106.63 ms]

ordered-builder-push-merge/100,000,000
time:   [1.1132 s 1.1171 s 1.1208 s]
```

**Comment by Kixiron** [comment]
Current nexmark benches show little to no change for me

Main:
```
┌───────┬─────────────┬───────┬─────────┬─────────────────┬──────────────────┬───────────────┬───────────────┬─────────────┐
│ Query │ #Events     │ Cores │ Elapsed │ Cores * Elapsed │ Throughput/Cores │ Total Usr CPU │ Total Sys CPU │ Max RSS(Kb) │
├───────┼─────────────┼───────┼─────────┼─────────────────┼──────────────────┼───────────────┼───────────────┼─────────────┤
│ q0    │ 100,000,000 │ 8     │ 14.492s │ 115.933s        │ 862.564 K/s      │ 125.390s      │ 0.375s        │ 127,620     │
│ q1    │ 100,000,000 │ 8     │ 14.516s │ 116.127s        │ 861.129 K/s      │ 139.375s      │ 0.188s        │ N/A         │
│ q2    │ 100,000,000 │ 8     │ 14.410s │ 115.280s        │ 867.452 K/s      │ 128.078s      │ 0.359s        │ N/A         │
│ q3    │ 100,000,000 │ 8     │ 14.713s │ 117.702s        │ 849.606 K/s      │ 130.782s      │ 0.406s        │ N/A         │
│ q4    │ 100,000,000 │ 8     │ 16.932s │ 135.458s        │ 738.236 K/s      │ 166.953s      │ 4.953s        │ N/A         │
│ q5    │ 100,000,000 │ 8     │ 14.407s │ 115.253s        │ 867.654 K/s      │ 133.078s      │ 0.844s        │ N/A         │
│ q6    │ 100,000,000 │ 8     │ 16.727s │ 133.813s        │ 747.311 K/s      │ 172.969s      │ 3.000s        │ N/A         │
│ q7    │ 100,000,000 │ 8     │ 20.739s │ 165.913s        │ 602.727 K/s      │ 205.328s      │ 5.313s        │ N/A         │
│ q8    │ 100,000,000 │ 8     │ 15.222s │ 121.780s        │ 821.153 K/s      │ 132.781s      │ 0.547s        │ N/A         │
│ q9    │ 100,000,000 │ 8     │ 74.303s │ 594.420s        │ 168.231 K/s      │ 346.422s      │ 5.296s        │ N/A         │
│ q12   │ 100,000,000 │ 8     │ 15.296s │ 122.367s        │ 817.211 K/s      │ 143.422s      │ 0.469s        │ N/A         │
│ q13   │ 100,000,000 │ 8     │ 21.014s │ 168.113s        │ 594.839 K/s      │ 206.468s      │ 3.344s        │ N/A         │
│ q14   │ 100,000,000 │ 8     │ 14.464s │ 115.708s        │ 864.242 K/s      │ 135.329s      │ 0.156s        │ N/A         │
│ q15   │ 100,000,000 │ 8     │ 15.275s │ 122.202s        │ 818.318 K/s      │ 167.437s      │ 2.063s        │ N/A         │
│ q16   │ 100,000,000 │ 8     │ 50.934s │ 407.470s        │ 245.417 K/s      │ 402.016s      │ 5.234s        │ N/A         │
│ q17   │ 100,000,000 │ 8     │ 21.841s │ 174.731s        │ 572.310 K/s      │ 247.672s      │ 3.766s        │ N/A         │
│ q18   │ 100,000,000 │ 8     │ 44.364s │ 354.912s        │ 281.760 K/s      │ 240.656s      │ 4.781s        │ N/A         │
│ q19   │ 100,000,000 │ 8     │ 44.184s │ 353.474s        │ 282.906 K/s      │ 327.969s      │ 12.328s       │ N/A         │
│ q20   │ 100,000,000 │ 8     │ 32.026s │ 256.206s        │ 390.311 K/s      │ 232.265s      │ 4.531s        │ N/A         │
│ q21   │ 100,000,000 │ 8     │ 15.673s │ 125.385s        │ 797.545 K/s      │ 143.610s      │ 0.344s        │ N/A         │
│ q22   │ 100,000,000 │ 8     │ 14.910s │ 119.281s        │ 838.358 K/s      │ 144.328s      │ 0.203s        │ N/A         │
└───────┴─────────────┴───────┴─────────┴─────────────────┴──────────────────┴───────────────┴───────────────┴─────────────┘
```

This Branch:
```
┌───────┬─────────────┬───────┬─────────┬─────────────────┬──────────────────┬───────────────┬───────────────┬─────────────┐
│ Query │ #Events     │ Cores │ Elapsed │ Cores * Elapsed │ Throughput/Cores │ Total Usr CPU │ Total Sys CPU │ Max RSS(Kb) │
├───────┼─────────────┼───────┼─────────┼─────────────────┼──────────────────┼───────────────┼───────────────┼─────────────┤
│ q0    │ 100,000,000 │ 8     │ 14.247s │ 113.980s        │ 877.349 K/s      │ 123.313s      │ 0.296s        │ 148,196     │
│ q1    │ 100,000,000 │ 8     │ 14.618s │ 116.941s        │ 855.129 K/s      │ 136.812s      │ 0.266s        │ N/A         │
│ q2    │ 100,000,000 │ 8     │ 15.008s │ 120.064s        │ 832.888 K/s      │ 127.828s      │ 0.328s        │ N/A         │
│ q3    │ 100,000,000 │ 8     │ 14.452s │ 115.613s        │ 864.955 K/s      │ 128.703s      │ 0.625s        │ N/A         │
│ q4    │ 100,000,000 │ 8     │ 16.834s │ 134.669s        │ 742.563 K/s      │ 168.219s      │ 4.922s        │ N/A         │
│ q5    │ 100,000,000 │ 8     │ 14.368s │ 114.943s        │ 869.998 K/s      │ 128.563s      │ 0.594s        │ N/A         │
│ q6    │ 100,000,000 │ 8     │ 16.808s │ 134.460s        │ 743.715 K/s      │ 179.203s      │ 3.265s        │ N/A         │
│ q7    │ 100,000,000 │ 8     │ 21.374s │ 170.990s        │ 584.829 K/s      │ 209.125s      │ 5.844s        │ N/A         │
│ q8    │ 100,000,000 │ 8     │ 15.061s │ 120.487s        │ 829.965 K/s      │ 132.547s      │ 0.360s        │ N/A         │
│ q9    │ 100,000,000 │ 8     │ 75.226s │ 601.810s        │ 166.165 K/s      │ 349.140s      │ 6.421s        │ N/A         │
│ q12   │ 100,000,000 │ 8     │ 15.426s │ 123.410s        │ 810.307 K/s      │ 141.282s      │ 0.454s        │ N/A         │
│ q13   │ 100,000,000 │ 8     │ 21.489s │ 171.915s        │ 581.683 K/s      │ 208.093s      │ 3.703s        │ N/A         │
│ q14   │ 100,000,000 │ 8     │ 14.723s │ 117.784s        │ 849.013 K/s      │ 133.750s      │ 0.265s        │ N/A         │
│ q15   │ 100,000,000 │ 8     │ 15.207s │ 121.659s        │ 821.972 K/s      │ 167.407s      │ 1.969s        │ N/A         │
│ q16   │ 100,000,000 │ 8     │ 50.642s │ 405.138s        │ 246.830 K/s      │ 410.265s      │ 4.000s        │ N/A         │
│ q17   │ 100,000,000 │ 8     │ 21.890s │ 175.116s        │ 571.048 K/s      │ 253.188s      │ 3.500s        │ N/A         │
│ q18   │ 100,000,000 │ 8     │ 43.835s │ 350.682s        │ 285.159 K/s      │ 245.703s      │ 5.719s        │ N/A         │
│ q19   │ 100,000,000 │ 8     │ 43.663s │ 349.301s        │ 286.286 K/s      │ 325.437s      │ 10.000s       │ N/A         │
│ q20   │ 100,000,000 │ 8     │ 32.045s │ 256.361s        │ 390.076 K/s      │ 231.469s      │ 4.047s        │ N/A         │
│ q21   │ 100,000,000 │ 8     │ 15.759s │ 126.076s        │ 793.174 K/s      │ 139.672s      │ 0.375s        │ N/A         │
│ q22   │ 100,000,000 │ 8     │ 14.044s │ 112.355s        │ 890.036 K/s      │ 132.281s      │ 0.312s        │ N/A         │
└───────┴─────────────┴───────┴─────────┴─────────────────┴──────────────────┴───────────────┴───────────────┴─────────────┘
```

**Comment by Kixiron** [comment]
I also enhanced nexmark query parsing somewhat, where previously invalid query names passed to `--query` would result in a silent failure (that is, using `--query something_invalid` would simply do nothing), it now knows what the available queries are and therefore can validate them, which leads to this behavior now:
```
λ  cargo bench --bench nexmark ... --query something_invalid
error: "something_invalid" isn't a valid value for '--query <QUERY>'
        [possible values: q0, q1, q2, q3, q4, q5, q6, q7, q8, q9, q12, q13, q14, q15, q16, q17, q18, q19, q20, q21, q22]

For more information try --help
error: bench failed, to rerun pass `--bench nexmark`
```
Additionally the help prompt now tells you the available queries:
```
λ  cargo bench --bench nexmark --help
...
        --query <QUERY>
            Queries to run, all by default

            [env: NEXMARK_QUERIES=]
            [possible values: q0, q1, q2, q3, q4, q5, q6, q7, q8, q9, q12, q13, q14, q15, q16, q17,
            q18, q19, q20, q21, q22]
...
```

**Comment by absoludity** [comment]
> I also enhanced nexmark query parsing somewhat, where previously invalid query names passed to `--query` would result in a silent failure (that is, using `--query something_invalid` would simply do nothing), it now knows what the available queries are and therefore can validate them, which leads to this behavior now
> 
> ```
> λ  cargo bench --bench nexmark ... --query something_invalid
> error: "something_invalid" isn't a valid value for '--query <QUERY>'
>         [possible values: q0, q1, q2, q3, q4, q5, q6, q7, q8, q9, q12, q13, q14, q15, q16, q17, q18, q19, q20, q21, q22]

Excellent, thanks @Kixiron .
 
> I also took advantage of using `mimalloc` to get more/richer/better stats on each query
> 

Nice, thanks! Yeah, I did look at using `mimalloc` when I saw you use it in a PR the other week, but I didn't spend enough time looking through your use of it to understand it well enough to use here.

From @ryzhyk's comment above, looks like he was just waiting for the performance numbers, which you've shown now to be minimally different?

**Comment by Kixiron** [comment]
Yah, me and Leon are still ironing out some performance oddities

**Pull request merged**